projects/ADuCM3029_demo_cn0428_cn0429/src/ADuCM3029_demo_cn0428_cn0429.cpp

```diff
@@ -0,0 +1,1307 @@
+	uint8_t 	cmdResponseBuff[256] = {0};	// buffer holding sensor response to commands
+	uint8_t		tempBuff[40] = {0};			// buffer for EIS results parsing
+	double		EISpartialresult[8] = {0};	// array of partial results (one line) of EIS
+	int32_t		EISpartialDFTresult[8] = {0};	// array of partial results (DFT impedance) of EIS
+	uint32_t	EISresults[192] = {0};		// array of EIS results
+	int16_t 	sigVal = 0;		// temporary variable for signed value
+	uint16_t 	unsigVal = 0;	// temporary variable for unsigned value
+	uint32_t	unsiglVal = 0;	// temporary variable for long unsigned value
+	char		tempChar = 0;   // temporary variable for char
+	bool 		exitloop = false;
+
+	uint8_t pulseDuration, pulseAmplitude;	// Pulse test parameters
+
+	uint8_t tempError;
+	uint8_t numbytes;
+	uint8_t cmdLen;
+	uint16_t timeout;
+
+	switch(FSM_State) {
+	case INIT:
+		FSM_State = COMMAND;
+
+		tempError = SensorInitwithI2CAddr(ADI_CFG_I2C_SENSOR1_ADDR, 1);
+		if (tempError == SENSOR_ERROR_NONE) {
+			S1_detected = true;
+		} else if (tempError == SENSOR_ERROR_PH) {	//Water Quality Board installed
+			S1_detected = true;
+			FSM_State = WATER;
+		}
+
+		tempError = SensorInitwithI2CAddr(ADI_CFG_I2C_SENSOR2_ADDR, 2);
+		if (tempError == SENSOR_ERROR_NONE) {
+			S2_detected = true;
+		} else if (tempError == SENSOR_ERROR_PH) {	//Water Quality Board installed
+			S2_detected = true;
+			FSM_State = WATER;
+		}
+
+		tempError = SensorInitwithI2CAddr(ADI_CFG_I2C_SENSOR3_ADDR, 3);
+		if (tempError == SENSOR_ERROR_NONE) {
+			S3_detected = true;
+		} else if (tempError == SENSOR_ERROR_PH) {	//Water Quality Board installed
+			S3_detected = true;
+			FSM_State = WATER;
+		}
+
+		tempError = SensorInitwithI2CAddr(ADI_CFG_I2C_SENSOR4_ADDR, 4);
+		if (tempError == SENSOR_ERROR_NONE) {
+			S4_detected = true;
+		} else if (tempError == SENSOR_ERROR_PH) {	//Water Quality Board installed
+			S4_detected = true;
+			FSM_State = WATER;
+		}
+
+		if(S1_detected||S2_detected||S3_detected||S4_detected) {
+			UART_TX("----------------------------------------------" _EOS);
+			cmdReceived = 1; //Ensure welcome message printout
+		} else UART_TX("No Sensor Board Found! Please check connections and reset ADICUP3029."
+				       _EOS);
+		break;
+
+	case COMMAND:
+		if (check_string((uint8_t *)"PRESENCE_CHECK", 14)) {
+			if (S1_detected) UART_TX("Sensor detected on site 1" _EOS);
+			if (S2_detected) UART_TX("Sensor detected on site 2" _EOS);
+			if (S3_detected) UART_TX("Sensor detected on site 3" _EOS);
+			if (S4_detected) UART_TX("Sensor detected on site 4" _EOS);
+		}
+
+		if (check_string((uint8_t *)"READ_CFGS", 9)) {
+			if (S1_detected) GetSensorCfg(ADI_CFG_I2C_SENSOR1_ADDR);
+			delay_ms(10);
+			if (S2_detected) GetSensorCfg(ADI_CFG_I2C_SENSOR2_ADDR);
+			delay_ms(10);
+			if (S3_detected) GetSensorCfg(ADI_CFG_I2C_SENSOR3_ADDR);
+			delay_ms(10);
+			if (S4_detected) GetSensorCfg(ADI_CFG_I2C_SENSOR4_ADDR);
+			delay_ms(10);
+		}
+
+		if (check_string((uint8_t *)"READ_TEMP", 9)) {
```

Comment:
these [check_string()] blocks also look repetitive;
they could also be grouped  into a more generic function/mode/style